chain/chain/src/store.rs

```diff
@@ -1717,111 +1752,255 @@ impl<'a> ChainStoreUpdate<'a> {
+            }
+
+            // 3. Delete chunk_hash-indexed data
+            // 3a. Delete chunks (ColChunks)
+            store_update.delete(ColChunks, chunk_header.hash.as_ref());
+            self.chain_store.chunks.cache_remove(&chunk_header.hash.clone().into());
+            // 3b. Delete chunk extras (ColChunkExtra)
+            store_update.delete(ColChunkExtra, chunk_header.hash.as_ref());
+            self.chain_store.chunk_extras.cache_remove(&chunk_header.hash.clone().into());
+            // 3c. Delete partial_chunks (ColPartialChunks)
+            store_update.delete(ColPartialChunks, chunk_header.hash.as_ref());
+            self.chain_store.partial_chunks.cache_remove(&chunk_header.hash.clone().into());
+            // 3d. Delete invalid chunks (ColInvalidChunks)
+            store_update.delete(ColInvalidChunks, chunk_header.hash.as_ref());
+            self.chain_store.invalid_chunks.cache_remove(&chunk_header.hash.clone().into());
+        }
+
+        // 4. Delete block_hash-indexed data
+        // 4a. Delete block (ColBlock) if not genesis
+        if height > 0 {
+            store_update.delete(ColBlock, block_hash.as_ref());
+            self.chain_store.blocks.cache_remove(&block_hash.clone().into());
+        }
+        // 4b. Delete block header (ColBlockHeader) - don't do because header sync needs headers
+        // 4c. Delete block extras (ColBlockExtra)
+        store_update.delete(ColBlockExtra, block_hash.as_ref());
+        self.chain_store.block_extras.cache_remove(&block_hash.clone().into());
+        // 4d. Delete from next_block_hashes (ColNextBlockHashes)
+        store_update.delete(ColNextBlockHashes, block_hash.as_ref());
+        self.chain_store.next_block_hashes.cache_remove(&block_hash.clone().into());
+        // 4e. Delete from my_last_approvals (ColMyLastApprovalsPerChain)
+        store_update.delete(ColMyLastApprovalsPerChain, block_hash.as_ref());
+        self.chain_store.my_last_approvals.cache_remove(&block_hash.clone().into());
+        // 4f. Delete from ColChallengedBlocks
+        store_update.delete(ColChallengedBlocks, block_hash.as_ref());
+        // 4g. Delete from ColBlocksToCatchup
+        store_update.delete(ColBlocksToCatchup, block_hash.as_ref());
+        // 4h. Delete from KV changes
+        let storage_key = block_hash.as_ref().to_vec();
+        let common_key_prefix_len = block_hash.as_ref().len();
+        // 4h1. We should collect all the keys which key prefix equals to `block_hash`
+        let changes_iter = self
+            .chain_store
+            .store()
+            .iter_prefix_ser::<RawStateChangesList>(ColKeyValueChanges, &storage_key);
+        // 4h2. Remove from ColKeyValueChanges all found KV changes
+        for change in changes_iter {
+            let (key, _) = change?;
+            store_update.delete(ColKeyValueChanges, &key[common_key_prefix_len..].to_owned());
+        }
```

Comment:
As of #2253 merged into master, here is how we should remove the state changes:

```rust
        let storage_key = KeyForStateChanges::get_prefix(&block_hash);
        // 4h1. We should collect all the keys which key prefix equals to `block_hash`
        let changes_iter = self
            .chain_store
            .store()
            .iter_prefix_ser::<Vec<u8>>(ColStateChanges, storage_key.as_ref());
        // 4h2. Remove from ColKeyValueChanges all found KV changes
        for change in changes_iter {
            let (key, _) = change?;
            store_update.delete(ColStateChanges, key);
        }
```

Notes:
* a type-safe `KeyForStateChanges` is used
* we don't need to deserialize into `RawStateChangesList`, so `Vec<u8>` is fine (ideally we would skip values completely, but I don't believe there is such an option).

Reply:
Awesome. Thank you for such detailed explanation.